api/src/context.rs

```diff
@@ -1022,7 +1028,77 @@ impl Context {
             }
         }
 
-        Ok((gas_prices, block_end_infos))
+        let majority_usecase_fraction = if count_majority_usecase {
+            count_by_usecase
+                .iter()
+                .max_by_key(|(_, v)| *v)
+                .and_then(|(max_usecase, max_value)| {
+                    if let UseCaseKey::ContractAddress(_) = max_usecase {
+                        Some(*max_value as f32 / count_by_usecase.values().sum::<u64>() as f32)
+                    } else {
+                        None
+                    }
+                })
+        } else {
+            None
+        };
+        Ok((gas_prices, block_end_infos, majority_usecase_fraction))
+    }
+
+    fn block_min_inclusion_price(
+        &self,
+        ledger_info: &LedgerInfo,
+        first: Version,
+        last: Version,
+        gas_estimation_config: &GasEstimationConfig,
+        execution_config: &OnChainExecutionConfig,
+    ) -> Option<u64> {
+        let user_use_case_spread_factor = execution_config
+            .transaction_shuffler_type()
+            .user_use_case_spread_factor();
+
+        match self.get_gas_prices_and_used(
+            first,
+            last - first,
+            ledger_info.ledger_version.0,
+            user_use_case_spread_factor.is_some(),
```

Comment:
Should we also have a local config here that can act as a kill switch if the gas estimates seem off?